src/v/cloud_storage/remote_partition.cc

```diff
@@ -75,115 +75,6 @@ remote_partition::iterator remote_partition::materialize_segment(
     return iter;
 }
 
-remote_partition::borrow_result_t remote_partition::borrow_next_segment_reader(
-
-  const partition_manifest& manifest,
-  storage::log_reader_config config,
-  model::offset hint) {
```

Comment:
The `borrow_next_reader` was implemented this way to avoid having to deal with scheduling points. The problem is that you have to check two collections which has to be somewhat consistent. Maybe. you can pre-acquire the units before calling this function and then pass the units to the `borrow_reader` instead of awaiting?

Reply:
The issue with pre-acquiring the units is that we don't know ahead of time if we actually need to acquire them. It's too pessimistic in my opinion. We may already have a materialised segment/partition reader and acquiring the units would hold up the read without a reason.

Reply:
It is possible to acquire optimistically and then, if the materialised segment exists the units could be returned. The code will only hold a unit for brief period of time which is probably fine.

Reply:
Happy to give it a go tomorrow. Sounds like it may be less intrusive.

Reply:
Done.